From: SeongJae Park <sj@kernel.org>
Date: Wed, 28 Jan 2026 17:07:48 -0800
Subject: [PATCH] mm/damon/core: align region size uppr limit by min_region_sz

Otherwise, regions of size not same to min_regions_sz could be made.

Worse yet, this generally could result int regions not aligned by
min_region_sz.  This is a real problem when online commit is triggered.
In the case, sysfs does three context commits.  First, generate a
user-input validity test-purpose blank context and commit the running
context's parameters to the blank context.  This invokes
damon_commit_target_regions() -> damon_set_regions().  The code reaches
the 'no region intersects with this range' part.  Here, it ALIGN_DOWN()
source start address and ALIGN() source end address and assign those to
the destination region.  If the addresses are not aligned by the
min_region_sz, and the source regions are contiguous, next region's
start address could be smaleer than the previous region's end address.
This means invalid regions are generated.  This is already sufficiently
bad.

The code further does second commit, which commits the user parameter to
the test-purpose context.  This time, damon_set_region() invoked and
'fill possible holes in the range' part is executed.  In the
damon_fill_regions_holes(), damon_new_region() with start address larger
than end address could be called, and damon_verify_new_region() triggers
BUG().  I'm not sure what real problem this will cause if we simply
ignore the BUG() by turning CONFIG_DAMON_HARDENED.  But, anyway, already
sufficiently bad.  And the fix is simple.  Fix it.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 656f19b629bc5..c0d144f73cae4 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -1544,8 +1544,9 @@ static unsigned long damon_region_sz_limit(struct damon_ctx *ctx)
 
 	if (ctx->attrs.min_nr_regions)
 		sz /= ctx->attrs.min_nr_regions;
-	if (sz < ctx->min_region_sz)
+	if (!sz)
 		sz = ctx->min_region_sz;
+	sz = ALIGN(sz, ctx->min_region_sz);
 
 	return sz;
 }
-- 
2.47.3

